compiler/noirc_evaluator/src/brillig/brillig_ir/codegen_control_flow.rs

```diff
@@ -182,115 +178,134 @@ impl<F: AcirField + DebugToString, Registers: RegisterAllocator> BrilligContext<
     }
 
     /// Jump to a trap condition if `condition` is false.
-    /// The trap will include the given message as revert data.
-    pub(crate) fn codegen_constrain_with_revert_data(
+    /// The trap will include the given message as error data.
+    ///
+    /// If `error_selector` is None, an empty error data is generated (no error information).
+    pub(crate) fn codegen_constrain_with_error_data(
         &mut self,
         condition: SingleAddrVariable,
-        revert_data_items: Vec<BrilligVariable>,
-        revert_data_types: Vec<BrilligParameter>,
-        error_selector: ErrorSelector,
+        error_data_items: Vec<BrilligVariable>,
+        error_data_types: Vec<BrilligParameter>,
+        error_selector: Option<ErrorSelector>,
     ) {
         assert!(condition.bit_size == 1);
 
         self.codegen_if_not(condition.address, |ctx| {
-            // + 1 due to the revert data id being the first item returned
-            let revert_data_size = Self::flattened_tuple_size(&revert_data_types) + 1;
-            let revert_data_size_var = ctx.make_usize_constant_instruction(revert_data_size.into());
-            let revert_data = ctx
+            let data_size = Self::flattened_tuple_size(&error_data_types);
+
+            // Special case: No error selector means completely empty error data
+            let Some(error_selector) = error_selector else {
+                let error_data =
+                    ctx.make_usize_constant_instruction(0_usize.into()).map(|size| HeapVector {
+                        pointer: ReservedRegisters::free_memory_pointer(),
+                        size: size.address,
+                    });
+                ctx.trap_instruction(*error_data);
+                return;
+            };
```

Comment:
This seems like we're moving logic which only exists for `codegen_constrain` into here whereas we could have them completely separate.

This also means that if we pass error data into here without an error selector then we silently ignore the error data.

Reply:
Yes it is indeed my point to move the logic of `codegen_constrain()` into `codegen_constrain_with_error_data()`, the latter being more general, and the former is now much simpler.

If we pass error data without an error selector, this check is precisely for this case and it uses the error data:  `ctx.trap_instruction(*error_data);`, so I am not sure I understand your point.